packages/oas/src/operation/lib/get-example-groups.ts

```diff
@@ -0,0 +1,207 @@
+       * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#example-object}
+       */
+      mediaTypeExample: MediaTypeExample;
+
+      /**
+       * The HTTP status code for the current response example
+       *
+       * @example "2xx"
+       * @example "400"
+       */
+      status: string;
+    };
+  }
+>;
+
+/**
+ * Internal key to represent custom code samples that do not have a corresponding response example.
+ */
+const noCorrespondingResponseKey = 'NoCorrespondingResponseForCustomCodeSample';
+
+/**
+ * Takes a groups object and an operation and adds any matching response examples
+ * to existing groups object
+ */
+function addMatchingResponseExamples(groups: ExampleGroups, operation: Operation) {
+  operation.getResponseExamples().forEach(example => {
+    Object.entries(example.mediaTypes || {}).forEach(([mediaType, mediaTypeExamples]) => {
+      mediaTypeExamples.forEach(mediaTypeExample => {
+        // only add a response example if the `title` field exists
+        // and it matches one of the existing example keys
+        if (mediaTypeExample.title && Object.keys(groups).includes(mediaTypeExample.title)) {
+          groups[mediaTypeExample.title].response = {
+            mediaType,
+            mediaTypeExample,
+            status: example.status,
+          };
+
+          // if the current group doesn't already have a name set,
+          // use the response example object's summary field
+          if (!groups[mediaTypeExample.title].name) {
+            groups[mediaTypeExample.title].name = mediaTypeExample.summary;
+          }
+        }
+      });
+    });
+  });
+}
+
+/**
+ * Returns a name for the given custom code sample. If there isn't already one defined,
+ * we construct a fallback value based on where the sample is in the array.
+ */
+function getDefaultName(sample: Extensions['code-samples'], count: number): string {
+  return sample.name && sample.name.length > 0 ? sample.name : `Default${count > 1 ? ` #${count}` : ''}`;
+}
+
+/**
+ * Returns an object with groups of all example definitions (body/header/query/path/response/etc.).
+ * The examples are grouped by their key when defined via the `examples` map.
+ *
+ * Any custom code samples defined via the `x-readme.code-samples` extension are returned,
+ * regardless of if they have a matching response example.
+ *
+ * For standard OAS request parameter (e.g., body/header/query/path/etc.) examples,
+ * they are only present in the return object if they have a corresponding response example
+ * (i.e., a response example with the same key in the `examples` map).
+ */
+export function getExampleGroups(operation: Operation): ExampleGroups {
+  let namelessCodeSamples = 0;
+  const groups: ExampleGroups = {};
+
+  // add custom code samples
+  const codeSamples = getExtension('code-samples', operation.api, operation) as Extensions['code-samples'][];
```

Comment:
Something for a future improvement but it'd be rad if `getExtension` could return the type based on the extension you specified so you could get rid of `as Extensions['code-samples'][]` here.

Might be able to accomplish it with type overrides? Would be neat.